src/MongoDB.Driver.Core/Core/Servers/Server.cs

```diff
@@ -988,6 +1002,20 @@ private async Task<TResult> ExecuteProtocolAsync<TResult>(IWireProtocol<TResult>
                     throw;
                 }
             }
+            
+            private async Task<TResult> ExecuteProtocolAsync<TResult>(IWireProtocol<TResult> protocol, ICoreSession session, CancellationToken cancellationToken)
+            {
+                try
+                {
+                    return await protocol.ExecuteAsync(_connection, cancellationToken).ConfigureAwait(false);
+                }
+                catch (Exception ex)
+                {
+                    MarkSessionDirtyIfNeeded(session, ex);
+                    _server.HandleChannelException(_connection, ex);
+                    throw;
+                }
+            }
```

Comment:
It is possible to amend overloaded method by specifying `ICoreSession session` as an optional parameter of  existing `ExecuteProtocolAsync<TResult>` method, but considering that `CancellationToken` is usually the last argument in method I would rather stick to overload option.

Reply:
I would say an optional parameter looks better here.

Reply:
Either way...